File: run_analysis.R
################################################################################
############################# run_analysis.R ###################################
### Constructs datasets for cell phone data from Samsung and export the data 
### into a csv file to be easily read in. This was part of a project for the 
### Coursera Getting and Cleaning Data course from JHU
################################################################################

# setwd("Your directory here if you want to specify it")

if (!require("data.table")) {
  install.packages("data.table")
  library(data.table)
}

construct_data<- function (string){
  # construct datasets based on string
  #
  # Args: 
  #   string: string indicating whether test or training set
  #
  # Returns:
  #   DT: a data.table of test or training data, depending on string entered
  df_file <- paste0('./project/UCI_HAR_Dataset/', string,
                    '/X_', string,'.txt')
    
  action_file <- paste0('./project/UCI_HAR_Dataset/', string, 
                        '/Y_', string, '.txt')
    
  subject_file <- paste0('./project/UCI_HAR_Dataset/', string, 
                         '/subject_', string, '.txt')
    
  DT <- fread(df_file, sep = ' ')
  DT$subject <- fread(subject_file, sep=' ')
  DT$action <- fread(action_file, sep=' ')
  setnames(DT, c(features, "subject", "action"))
    
  return (DT)
}

# Create the project directory
if (!dir.exists('project')){dir.create('project')}

#if directory empty, download zip file and unzip it
if (length(dir('./project')) == 0){
    
  projectURL <- paste0('https://d396qusza40orc.cloudfront.net/',
                       'getdata%2Fprojectfiles%2FUCI%20HAR%20Dataset.zip')
  download.file(projectURL, destfile= './project/project.zip', 
                method='curl')
  datedownloaded = date()

  # unzip the contents & download the dataset.
  unzip('./project/project.zip', list=FALSE, exdir = './project')

  # rename file
  file.rename('./project/UCI HAR Dataset', './project/UCI_HAR_Dataset')
}

# Uncomment to see what's in file
# list.files('./project/UCI_HAR_Dataset')

# pull in the features.txt info for labels
features <- read.table("./project/UCI_HAR_Dataset/features.txt", 
                       sep = " ", colClasses = "character")[,2]

# construct each dataset 
testDT <- construct_data('test')
trainDT <- construct_data('train')

# append the datasets on proper variables
appendedDT <- rbindlist(list(testDT, trainDT))

rm(testDT, trainDT, features)

# clean up the dt_names
dt_names <- grep('mean|std', names(appendedDT), 
                 value=TRUE, ignore.case = TRUE)
dt_names <- append(dt_names, c('subject', 'action'))

# use dt_names to select those variables, switch to data.table notation 
appendedDT <- appendedDT[, dt_names, with=FALSE]

rm(dt_names)

# use activity_labels.txt to match numbers in dataset to the activities 
activity <- read.table('./project/UCI_HAR_Dataset/activity_labels.txt',
                       sep='', col.names = c("action", "activity_performed"))

mergedDT <- merge(appendedDT, activity, by='action')

rm(activity, appendedDT)

# Loop through and replace various components of the variable names rather than
# perform them manually. 
  
merged_names <- names(mergedDT)

keywords <- c('^t', '^f', '\\.*X', '\\.*Y', '\\.*Z', '\\.', 'mean','std','-')
strings <- c('time', 'freq', 'X', 'Y', 'Z', '', 'Mean', 'StdDev', '_')

for (i in 1:length(keywords)){
    merged_names <- gsub(keywords[i], strings[i], merged_names)
}

# change original names of DT (double check this)
setnames(mergedDT, merged_names) 

# group by activity and subject, and report the mean for each variable. 
DT <- mergedDT[, lapply(.SD, mean), by=c('subject', 'action')]

# Finally, write this dataset out into a .txt file. 
write.table(DT, "MeanData.txt", sep=' ', row.names = FALSE)


# remove objects in memory
rm(list = ls())

# if you want to delete the project folder after running, 
# uncomment the following line:
# unlink('./project', recursive = TRUE)

##############################################################################

File: run_analysis.py
import os
import pandas as pd
import numpy as np
#from numpy import loadtext
import requests, zipfile, StringIO, re
os.getcwd()
home = os.path.expanduser("~")
os.chdir(home+"/Desktop/Statistical_Programming/"+
            "Coursera/Getting_And_Cleaning_Data/Project_Three")
#dir = 'project'

URL = 'https://d396qusza40orc.cloudfront.net/' +\
    'getdata%2Fprojectfiles%2FUCI%20HAR%20Dataset.zip'

#create directory if it doesn't exist
if not os.path.exists('./project'):
    os.makedirs('project')

os.listdir("./")

#if project folder empty, download file
if not os.path.exists('./project/UCI_HAR_Dataset'):

    #download zip and write to disk
    print "This download will take a while..."
    response = requests.get(URL)
    with open('./project/project.zip','wb') as fout:
        fout.write(response.content)

    #get the download date
    date = response.headers['Date']
    del(response)

    #unzip
    zf = zipfile.ZipFile('./project/project.zip')
    zf.extractall('./project')


    #rename the unzipped file
    os.rename('./project/UCI HAR Dataset',
        './project/UCI_HAR_Dataset')

#see what's in the file
#os.listdir('./project')

####################Step 1: Combine Datasets########################################
features = pd.read_table('./project/UCI_HAR_Dataset/features.txt',
                header = None, delim_whitespace=True,
                usecols=[1], squeeze=True)

def construct_data(string):
    df_file = './project/UCI_HAR_Dataset/'+ string+'/X_'+ \
                    string+'.txt'
    action_file = './project/UCI_HAR_Dataset/'+string+'/Y_'+ \
                        string+'.txt'
    subject_file = './project/UCI_HAR_Dataset/'+string+'/subject_'+\
                        string+'.txt'

    df = pd.read_table(df_file, delim_whitespace=True,
                    header = None, names = features)
    df['subject'] = pd.read_table(subject_file, delim_whitespace=True,
                    squeeze=True,
                    header = None, names=['subject'])
    df['action'] = pd.read_table(action_file, delim_whitespace=True,
                    squeeze=True,
                    header = None, names=['action'])

    return df

test_df = construct_data('test')
train_df = construct_data('train')

appendedDF = test_df.append(train_df)

#remove unnecessary objects from memory
del(test_df, train_df)

####################Step 2: Grab only mean/std. dev data############################

# try to restructure using lambda function perhaps, or list comprehension 
# feature_test = filter(lambda x:
#                 re.search(r'[*mean|*std|\bsubject\b|\baction\b]', x),
#                     features)
# feature_test = filter(lambda x: re.search(r'[mean]', x), features)
    
feature_test = []
for item in features:
    match = re.search(r"[M|m]ean|[S|s]td", item)
    if match:
        feature_test.append(item)
        
features = feature_test
features.append('action')
features.append('subject')
appendedDF = appendedDF[features]

####################Step 3: assign labels and subjects to data######################
activity = pd.read_table('./project/UCI_HAR_Dataset/activity_labels.txt',
                names = ["action","activity_performed"], header = None, 
                delim_whitespace=True,squeeze=True)

mergedDF = pd.merge(appendedDF, activity,  on='action', how='inner')

del(activity, appendedDF)

####################Step 4: Clean up names in data##################################
merged_names = mergedDF.columns.values

keywords = ['^t', '^f', '\\.*X', '\\.*Y', '\\.*Z', '\\.', 'mean', 'std', '-']
strings  = ['time', 'freq', 'X', 'Y', 'Z', '', 'Mean', 'StdDev', '_']

for i in xrange(len(merged_names)):
    for j in xrange(len(keywords)):
        merged_names[i] = re.sub(keywords[j], strings[j], merged_names[i])

mergedDF.columns = merged_names

####################Step 5: Get dataset of means of each var########################
#group by activity and subject, and report the mean for each variable. 
meanDF = mergedDF.groupby(["subject", 'action']).mean()


###################Write out a text file of data####################################
meanDF.to_csv(r'./project/MeanData.txt', header= meanDF.columns.values, \
                index=None, sep=' ', mode='w')

####################Remove Objects from memory######################################
for name in dir():
    if not name.startswith('_'):
        del globals()[name]

#if you want to delete the project folder after running, 
#uncomment the following line:
#fill with command to delete contents and folder in python 

##############################################################################